File: solana-wallet-checker/csv_importer.py
"""
CSV Importer Module for Solana Wallet Checker Bot.
Imports holder data from Solscan CSV exports.
"""

import csv
from typing import List, Dict, Optional
from datetime import datetime


class CSVImporter:
    """Imports and parses holder data from CSV files."""
    
    def __init__(self):
        """Initialize CSV importer."""
        pass
    
    def parse_csv(self, file_path: str, token_mint: str = None) -> Dict:
        """
        Parse CSV file containing holder data.
        
        Supports Solscan CSV format with columns:
        - Rank, Address, Quantity, Percentage, Value, etc.
        
        Args:
            file_path: Path to CSV file
            token_mint: Optional token mint address for metadata
            
        Returns:
            Dictionary with holders list and metadata
        """
        holders = []
        
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                # Try to detect CSV format
                sample = f.read(1024)
                f.seek(0)
                
                # Auto-detect delimiter
                sniffer = csv.Sniffer()
                try:
                    delimiter = sniffer.sniff(sample).delimiter
                except:
                    delimiter = ','
                
                reader = csv.DictReader(f, delimiter=delimiter)
                
                for row in reader:
                    # Parse holder data - support multiple CSV formats
                    holder = self._parse_row(row)
                    if holder:
                        holders.append(holder)
            
            if not holders:
                raise ValueError("No valid holder data found in CSV")
            
            # Calculate total balance
            total_balance = sum(h['balance'] for h in holders)
            
            # Add percentage if not present
            for holder in holders:
                if 'percentage' not in holder or holder['percentage'] == 0:
                    holder['percentage'] = (holder['balance'] / total_balance * 100) if total_balance > 0 else 0
            
            return {
                'holders': holders,
                'total_holders': len(holders),
                'total_balance': total_balance,
                'token_mint': token_mint,
                'import_time': datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            }
            
        except FileNotFoundError:
            raise FileNotFoundError(f"CSV file not found: {file_path}")
        except Exception as e:
            raise RuntimeError(f"Error parsing CSV: {e}")
    
    def _parse_row(self, row: Dict[str, str]) -> Optional[Dict]:
        """
        Parse a single CSV row into holder data.
        
        Supports multiple column name formats:
        - Solscan: "Rank", "Address", "Quantity", "Percentage"
        - Generic: "rank", "address", "balance", "amount", "percent"
        
        Args:
            row: Dictionary from CSV row
            
        Returns:
            Holder dictionary or None if invalid
        """
        # Normalize column names (case-insensitive)
        row_lower = {k.lower().strip(): v.strip() for k, v in row.items()}
        
        # Extract address (required)
        address = None
        for key in ['address', 'owner', 'wallet', 'account']:
            if key in row_lower and row_lower[key]:
                address = row_lower[key]
                break
        
        if not address or len(address) < 32:
            return None
        
        # Extract balance/quantity (required)
        balance = 0.0
        for key in ['quantity', 'balance', 'amount', 'tokens']:
            if key in row_lower and row_lower[key]:
                try:
                    # Remove commas and convert to float
                    balance_str = row_lower[key].replace(',', '').replace(' ', '')
                    balance = float(balance_str)
                    break
                except (ValueError, AttributeError):
                    continue
        
        if balance <= 0:
            return None
        
        # Extract percentage (optional)
        percentage = 0.0
        for key in ['percentage', 'percent', '%', 'share']:
            if key in row_lower and row_lower[key]:
                try:
                    # Remove % sign and convert
                    percent_str = row_lower[key].replace('%', '').replace(',', '').strip()
                    percentage = float(percent_str)
                    break
                except (ValueError, AttributeError):
                    continue
        
        # Extract rank (optional)
        rank = None
        for key in ['rank', '#', 'no', 'number']:
            if key in row_lower and row_lower[key]:
                try:
                    rank = int(row_lower[key])
                    break
                except (ValueError, AttributeError):
                    continue
        
        return {
            'owner': address,
            'balance': balance,
            'percentage': percentage,
            'rank': rank,
            'token_account': None,  # Not available from CSV
            'purchase_time': None,
            'purchase_time_str': 'Unknown',
            'token_count': 0  # Will be filled by similarity analysis
        }
    
    def validate_csv_format(self, file_path: str) -> Dict[str, any]:
        """
        Validate CSV file format and return info.
        
        Args:
            file_path: Path to CSV file
            
        Returns:
            Dictionary with validation results
        """
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                # Read first few lines
                sample = f.read(2048)
                f.seek(0)
                
                reader = csv.DictReader(f)
                headers = reader.fieldnames
                
                # Count rows
                row_count = sum(1 for _ in reader)
                
                # Check for required columns
                headers_lower = [h.lower() for h in headers]
                has_address = any(k in headers_lower for k in ['address', 'owner', 'wallet', 'account'])
                has_balance = any(k in headers_lower for k in ['quantity', 'balance', 'amount', 'tokens'])
                
                return {
                    'valid': has_address and has_balance,
                    'row_count': row_count,
                    'headers': headers,
                    'has_address': has_address,
                    'has_balance': has_balance,
                    'sample': sample[:500]
                }
                
        except Exception as e:
            return {
                'valid': False,
                'error': str(e)
            }

File: solana-wallet-checker-js/src/csvImporter.js
/**
 * CSV Importer Module for Solana Wallet Checker Bot.
 * Imports holder data from Solscan CSV exports.
 * 
 * Migrated from Python to Node.js with csv-parse
 */

import { createReadStream, readFileSync } from 'fs';
import { parse } from 'csv-parse/sync';

export class CSVImporter {
  constructor() {}

  /**
   * Parse CSV file containing holder data.
   * Supports Solscan CSV format with columns: Rank, Address, Quantity, Percentage, etc.
   * 
   * @param {string} filePath - Path to CSV file
   * @param {string|null} tokenMint - Optional token mint address for metadata
   * @returns {object}
   */
  parseCSV(filePath, tokenMint = null) {
    try {
      const content = readFileSync(filePath, 'utf-8');

      // Auto-detect delimiter
      let delimiter = ',';
      const firstLine = content.split('\n')[0];
      if (firstLine.split('\t').length > firstLine.split(',').length) {
        delimiter = '\t';
      } else if (firstLine.split(';').length > firstLine.split(',').length) {
        delimiter = ';';
      }

      const records = parse(content, {
        columns: true,
        skip_empty_lines: true,
        delimiter,
        trim: true,
        relax_column_count: true,
      });

      const holders = [];
      for (const row of records) {
        const holder = this._parseRow(row);
        if (holder) holders.push(holder);
      }

      if (holders.length === 0) {
        throw new Error('No valid holder data found in CSV');
      }

      // Calculate total balance
      const totalBalance = holders.reduce((sum, h) => sum + h.balance, 0);

      // Add percentage if not present
      for (const holder of holders) {
        if (!holder.percentage || holder.percentage === 0) {
          holder.percentage = totalBalance > 0
            ? (holder.balance / totalBalance * 100)
            : 0;
        }
      }

      return {
        holders,
        totalHolders: holders.length,
        totalBalance,
        tokenMint,
        importTime: new Date().toISOString().replace('T', ' ').split('.')[0],
      };
    } catch (err) {
      if (err.code === 'ENOENT') {
        throw new Error(`CSV file not found: ${filePath}`);
      }
      throw new Error(`Error parsing CSV: ${err.message}`);
    }
  }

  /**
   * Parse a single CSV row into holder data.
   * @param {object} row 
   * @returns {object|null}
   */
  _parseRow(row) {
    // Normalize column names (case-insensitive)
    const rowLower = {};
    for (const [key, value] of Object.entries(row)) {
      rowLower[key.toLowerCase().trim()] = (value || '').trim();
    }

    // Extract address (required)
    let address = null;
    for (const key of ['address', 'owner', 'wallet', 'account']) {
      if (rowLower[key]) {
        address = rowLower[key];
        break;
      }
    }
    if (!address || address.length < 32) return null;

    // Extract balance/quantity (required)
    let balance = 0;
    for (const key of ['quantity', 'balance', 'amount', 'tokens']) {
      if (rowLower[key]) {
        try {
          balance = parseFloat(rowLower[key].replace(/,/g, '').replace(/ /g, ''));
          if (!isNaN(balance) && balance > 0) break;
        } catch (e) {
          continue;
        }
      }
    }
    if (balance <= 0) return null;

    // Extract percentage (optional)
    let percentage = 0;
    for (const key of ['percentage', 'percent', '%', 'share']) {
      if (rowLower[key]) {
        try {
          percentage = parseFloat(rowLower[key].replace(/%/g, '').replace(/,/g, '').trim());
          if (!isNaN(percentage)) break;
        } catch (e) {
          continue;
        }
      }
    }

    // Extract rank (optional)
    let rank = null;
    for (const key of ['rank', '#', 'no', 'number']) {
      if (rowLower[key]) {
        try {
          rank = parseInt(rowLower[key], 10);
          if (!isNaN(rank)) break;
        } catch (e) {
          continue;
        }
      }
    }

    return {
      owner: address,
      balance,
      percentage: percentage || 0,
      rank,
      tokenAccount: null,
      purchaseTime: null,
      purchaseTimeStr: 'Unknown',
      tokenCount: 0,
    };
  }

  /**
   * Validate CSV file format and return info.
   * @param {string} filePath 
   * @returns {object}
   */
  validateCSVFormat(filePath) {
    try {
      const content = readFileSync(filePath, 'utf-8');
      const lines = content.split('\n').filter(l => l.trim());
      
      if (lines.length < 2) {
        return { valid: false, error: 'File has no data rows' };
      }

      // Parse just the header
      const records = parse(content, {
        columns: true,
        skip_empty_lines: true,
        trim: true,
        relax_column_count: true,
        to: 1, // Only read 1 data row to validate
      });

      const headers = Object.keys(records[0] || {});
      const headersLower = headers.map(h => h.toLowerCase());

      const hasAddress = ['address', 'owner', 'wallet', 'account'].some(k => headersLower.includes(k));
      const hasBalance = ['quantity', 'balance', 'amount', 'tokens'].some(k => headersLower.includes(k));

      return {
        valid: hasAddress && hasBalance,
        rowCount: lines.length - 1, // Exclude header
        headers,
        hasAddress,
        hasBalance,
        sample: content.slice(0, 500),
      };
    } catch (err) {
      return { valid: false, error: err.message };
    }
  }
}
